src/protocol/frame/init_aware_buf.rs

```diff
@@ -0,0 +1,221 @@
+    pub fn len(&self) -> usize {
+        self.bytes.len()
+    }
+
+    #[inline]
+    pub fn capacity(&self) -> usize {
+        self.bytes.capacity()
+    }
+
+    #[inline]
+    pub fn split_to(&mut self, at: usize) -> BytesMut {
+        let split = self.bytes.split_to(at);
+        self.init_cap -= at;
+        split
+    }
+
+    #[inline]
+    pub fn reserve(&mut self, additional: usize) {
+        // Increasing capacity doesn't change `init_cap`
+        self.bytes.reserve(additional);
+    }
+
+    /// Sets the length of the buffer to `len`. If above the current
+    /// initialized capacity any uninitialized bytes will be zeroed.
+    ///
+    /// This is more efficient that [`BytesMut::resize`] as spare capacity
+    /// is only initialized **once** past the initialized_capacity. This
+    /// allow the method to be efficiently called after truncating.
+    ///
+    /// # Panics
+    /// Panics if `len > capacity`.
+    #[inline]
+    pub fn resize(&mut self, len: usize) {
+        if len <= self.init_cap {
+            // SAFETY: init_cap tracks initialised bytes.
+            unsafe {
+                self.bytes.set_len(len);
+            }
+        } else {
+            assert!(len <= self.capacity());
+            let cur_len = self.bytes.len();
+            let spare = self.bytes.spare_capacity_mut();
+            let already_init = self.init_cap - cur_len;
+            let zeroes = len - self.init_cap;
+            debug_assert!(already_init + zeroes <= spare.len());
+            unsafe {
+                // SAFETY: spare capacity is sufficient for `zeroes` extra bytes
+                ptr::write_bytes(spare[already_init..].as_mut_ptr().cast::<u8>(), 0, zeroes);
+                // SAFETY: len has been initialized
+                self.bytes.set_len(len);
+            }
+            self.init_cap = len;
+        }
+    }
+
+    #[inline]
+    pub fn truncate(&mut self, len: usize) {
+        // truncating doesn't change `init_cap`
+        self.bytes.truncate(len);
+    }
+
+    #[inline]
+    pub fn advance(&mut self, cnt: usize) {
+        self.bytes.advance(cnt);
+        self.init_cap -= cnt;
+    }
+}
+
+impl From<BytesMut> for InitAwareBuf {
+    #[inline]
+    fn from(bytes: BytesMut) -> Self {
+        let init_cap = bytes.len();
+        Self { bytes, init_cap }
+    }
+}
+
+impl From<InitAwareBuf> for BytesMut {
+    #[inline]
+    fn from(value: InitAwareBuf) -> Self {
+        value.bytes
+    }
+}
+
+impl AsRef<[u8]> for InitAwareBuf {
+    #[inline]
+    fn as_ref(&self) -> &[u8] {
+        &self.bytes
+    }
+}
+
+impl Deref for InitAwareBuf {
+    type Target = [u8];
+
+    #[inline]
+    fn deref(&self) -> &[u8] {
+        &self.bytes
+    }
+}
+
+impl AsMut<[u8]> for InitAwareBuf {
+    #[inline]
+    fn as_mut(&mut self) -> &mut [u8] {
+        &mut self.bytes
+    }
+}
+
+impl DerefMut for InitAwareBuf {
+    #[inline]
+    fn deref_mut(&mut self) -> &mut [u8] {
+        &mut self.bytes
+    }
+}
```

Comment:
I come from the `bytes` issue. It's taking me some time to review this because of the manual slicing that happens in the other files. I'm not a fan of them, because in a way they leak implementation internals and leave the other modules to do the slicing by themselves. Although it's an internal API, it doesn't feel robust.

Why not copy the `read_buf` API (both the initial one which you can see in tokio, and the current one) by having methods like:

```rs
impl InitAwareBuf {
    pub fn filled(&self) -> &[u8] {}

    pub fn init_mut(&mut self) -> &mut [u8] {}

    pub fn advance_mut(&mut self, filled_len: usize) {}
}
```

Reply:
I envisioned the wrapper as transparent bytes but with extra info, the initialised capacity. Slice access, even mut, is fine as it doesn't grow or shrink the buffer.

This style simplifies the overall change as usage of the buf, previously plain `BytesMut` is fairly unchanged.

We just need ensure the new wrapper itself is sound.